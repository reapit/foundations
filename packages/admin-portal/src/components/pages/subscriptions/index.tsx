import React, { useEffect } from 'react'
import { History } from 'history'
import { useSelector } from 'react-redux'
import { useHistory, useLocation } from 'react-router'
import ErrorBoundary from '@/components/hocs/error-boundary'
import {
  Pagination,
  Table,
  Loader,
  setQueryParams,
  H3,
  toLocalTime,
  isEmptyObject,
  Section,
  Alert,
  Button,
  notification,
} from '@reapit/elements'
import Routes from '@/constants/routes'
import SubscriptionsFilterForm, {
  SubscriptionsFilterFormValues,
} from '@/components/ui/subscriptions/subscriptions-filter-form'
import CancelConfirmModal from '@/components/ui/subscriptions/subscription-cancel-confirm'
import MemberNameCell from '@/components/ui/subscriptions/subscription-member-name-cell'
import { selectSubscriptionListState } from '@/selector/admin'
import { cleanObject, errorMessages } from '@reapit/utils'
import {
  fetchSubscriptionList,
  FetchSubscriptionListQuery,
  cancelSubscription as cancelSubscriptionAc,
} from '@/actions/subscriptions'
import { CancelSubscriptionParams } from '@/services/subscriptions'
import store from '../../../core/store'

export const buildFilterValues = (queryParams: URLSearchParams): SubscriptionsFilterFormValues => {
  const type = queryParams.get('type') || ''
  const developerId = queryParams.get('developerId') || ''
  return { type, developerId } as SubscriptionsFilterFormValues
}

export const onPageChangeHandler = (history: History<any>, queryParams: SubscriptionsFilterFormValues) => (
  page: number,
) => {
  const query = setQueryParams(queryParams)
  let queryString = `?page=${page}`
  if (query && query !== '') {
    queryString = queryString.concat(`&${query}`)
  }
  return history.push(`${Routes.SUBSCRIPTIONS}${queryString}`)
}

export const onSearchHandler = (history: History<any>) => (
  queryParams: SubscriptionsFilterFormValues,
  { setStatus },
) => {
  const cleanedValues = cleanObject(queryParams)
  const { developerId } = cleanedValues
  if (developerId?.length > 1) {
    return notification.error({
      message: errorMessages.SUBSCRIPTION_MULTIPLE_DEVELOPER,
      placement: 'bottomRight',
    })
  }

  if (isEmptyObject(cleanedValues)) {
    setStatus('Please enter at least one search criteria')
    return
  }

  setStatus(null)
  const query = setQueryParams(cleanedValues)
  if (query && query !== '') {
    const queryString = `?page=1&${query}`
    history.push(`${Routes.SUBSCRIPTIONS}${queryString}`)
  }
}

const Subscriptions: React.FC = () => {
  const history = useHistory()
  const location = useLocation()
  const search = location.search
  const queryParams = new URLSearchParams(location.search)
  const filterValues = buildFilterValues(queryParams)
  const onPageChange = React.useCallback(onPageChangeHandler(history, filterValues), [history, filterValues])
  const onSearch = React.useCallback(onSearchHandler(history), [history])

  const SubscriptionsListState = useSelector(selectSubscriptionListState)

  const page = queryParams.get('page') ? Number(queryParams.get('page')) : 1
  useEffect(() => {
    if (search) {
      store.dispatch(fetchSubscriptionList({ page, queryString: search } as FetchSubscriptionListQuery))
    }
  }, [search])

  const [cancelSubId, setCancelSubId] = React.useState<string>('')

  const onCancelSubscription = () => {
    store.dispatch(
      cancelSubscriptionAc({
        id: cancelSubId,
      } as CancelSubscriptionParams),
    )
  }

  const { data, totalCount, pageSize, pageNumber = 1, isLoading, cancelSubscription } = SubscriptionsListState

  const CreatedCell = ({ cell: { value } }) => <p>{toLocalTime(value)}</p>

  const StatusCell = ({ cell: { value } }) => <p>{value ? 'Deactived' : 'Active'}</p>

  const DeleteButton = ({ cell: { value } }) => (
    <Button type="button" variant="danger" onClick={() => setCancelSubId(value)}>
      Cancel
    </Button>
  )

  const columns = [
    { Header: 'Subscription Type', accessor: 'type' },
    { Header: 'Summary', accessor: 'summary' },
    { Header: 'Application ID', accessor: 'applicationId' },
    { Header: 'Member Name', accessor: 'developerId', Cell: MemberNameCell },
    { Header: 'Member Email', accessor: 'user' },
    {
      Header: 'Start Date',
      accessor: 'created',
      Cell: CreatedCell,
    },
    {
      Header: 'Renewal Date',
      accessor: 'renews',
    },
    { Header: 'Frequency', accessor: 'frequency' },
    { Header: 'Cost', accessor: 'cost' },
    { Header: 'Status', accessor: 'cancelled', Cell: StatusCell },
    {
      Header: '',
      accessor: 'id',
      Cell: DeleteButton,
    },
  ]

  return (
    <ErrorBoundary>
      <H3 isHeadingSection>Subscriptions</H3>
      <SubscriptionsFilterForm filterValues={filterValues} onSearch={onSearch} />
      {isLoading || !data ? (
        <Loader />
      ) : (
        <>
          {renderResult(data, columns, totalCount)}
          <Pagination onChange={onPageChange} totalCount={totalCount} pageSize={pageSize} pageNumber={pageNumber} />
        </>
      )}
      <CancelConfirmModal
        onConfirm={onCancelSubscription}
        isShowConfirmModal={cancelSubId !== ''}
        setCancelSubId={setCancelSubId}
        isCanceling={cancelSubscription.isLoading}
      />
    </ErrorBoundary>
  )
}

export const renderResult = (data, columns, totalCount) => {
  if (data?.length === 0) {
    return <Alert message="No Results " type="info" />
  }

  return (
    <>
      <Section>
        <div>Total: {totalCount}</div>
      </Section>
      <Section>
        <Table expandable scrollable={true} loading={false} data={data || []} columns={columns} />
      </Section>
    </>
  )
}

export default Subscriptions
