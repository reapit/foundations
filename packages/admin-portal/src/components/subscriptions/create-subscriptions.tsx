import React, { useEffect, Dispatch, SetStateAction, useState, FC } from 'react'
import {
  CreateSubscriptionModel,
  SubscriptionModel,
  SubscriptionModelPagedResult,
} from '@reapit/foundations-ts-definitions'
import { objectToQuery, SendFunction, useReapitGet, useReapitUpdate } from '@reapit/utils-react'
import { GetActionNames, getActions, UpdateActionNames, updateActions } from '@reapit/utils-common'
import { Button } from '@reapit/elements'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { useReapitConnect } from '@reapit/connect-session'
import { usePermissionsState } from '../../core/use-permissions-state'

export type SubscriptionType = 'applicationListing' | 'developerRegistration' | 'developerEdition'

export interface CreateSubscriptionsButtonProps {
  subscriptionType: SubscriptionType
  developerId?: string
  appId?: string
}

export const getCurrentSub =
  (
    subscriptions: SubscriptionModelPagedResult | null,
    subscriptionType: SubscriptionType,
    setCurrentSub: Dispatch<SetStateAction<SubscriptionModel | null>>,
  ) =>
  () => {
    if (subscriptions?.data && subscriptionType === 'applicationListing') {
      const sub = subscriptions.data.find((sub) => sub.type === 'applicationListing' && !sub.cancelled) ?? null
      setCurrentSub(sub)
    }

    if (subscriptions?.data && subscriptionType === 'developerRegistration') {
      const sub = subscriptions.data.find((sub) => sub.type === 'developerRegistration' && !sub.cancelled) ?? null
      setCurrentSub(sub)
    }
  }

export const createSubscriptionHander =
  (
    createSubscription: SendFunction<CreateSubscriptionModel, boolean>,
    createSubscriptionModel: CreateSubscriptionModel,
  ) =>
  () => {
    createSubscription(createSubscriptionModel)
  }

export const cancelSubscriptionHander = (cancelSubscription: SendFunction<void, boolean>) => () => {
  cancelSubscription()
}

export const handleRefreshSubs = (subscriptionsRefresh: () => void, shouldRefresh?: boolean) => () => {
  if (shouldRefresh) {
    subscriptionsRefresh()
  }
}

export const handleFetchSubs = (setShouldFetchSubs: Dispatch<SetStateAction<boolean>>) => () => {
  setShouldFetchSubs(true)
}

export const CreateSubscriptionsButton: FC<CreateSubscriptionsButtonProps> = ({
  subscriptionType,
  developerId,
  appId,
}) => {
  const { connectSession } = useReapitConnect(reapitConnectBrowserSession)
  const { hasReadAccess } = usePermissionsState()
  const email = connectSession?.loginIdentity.email
  const [currentSub, setCurrentSub] = useState<SubscriptionModel | null>(null)
  const [shouldFetchSubs, setShouldFetchSubs] = useState<boolean>(false)
  const queryParams = objectToQuery({
    appId,
    developerId,
    subscriptionType,
  })

  const [subscriptions, subscriptionsLoading, , subscriptionsRefresh] = useReapitGet<SubscriptionModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions(window.reapit.config.appEnv)[GetActionNames.getSubscriptions],
    queryParams: {
      ...queryParams,
      pageSize: 999,
    },
    fetchWhenTrue: [shouldFetchSubs],
  })

  const [cancelSubscriptionLoading, , cancelSubscription, subscriptionCancelled] = useReapitUpdate<void, boolean>({
    reapitConnectBrowserSession,
    action: updateActions(window.reapit.config.appEnv)[UpdateActionNames.deleteSubscription],
    method: 'DELETE',
    uriParams: {
      subscriptionId: currentSub?.id,
    },
  })

  const [createSubscriptionLoading, , createSubscription, subscriptionCreated] = useReapitUpdate<
    CreateSubscriptionModel,
    boolean
  >({
    reapitConnectBrowserSession,
    action: updateActions(window.reapit.config.appEnv)[UpdateActionNames.createSubscription],
    method: 'POST',
  })

  const shouldRefresh = subscriptionCancelled || subscriptionCreated

  useEffect(getCurrentSub(subscriptions, subscriptionType, setCurrentSub), [subscriptions, subscriptionType])
  useEffect(handleRefreshSubs(subscriptionsRefresh, shouldRefresh), [subscriptionCancelled, subscriptionCreated])

  return !shouldFetchSubs || !subscriptions ? (
    <Button
      onClick={handleFetchSubs(setShouldFetchSubs)}
      intent="secondary"
      loading={subscriptionsLoading}
      disabled={hasReadAccess || subscriptionsLoading}
    >
      Check Subscriptions
    </Button>
  ) : currentSub ? (
    <Button
      onClick={cancelSubscriptionHander(cancelSubscription)}
      intent="secondary"
      loading={cancelSubscriptionLoading || subscriptionsLoading}
      disabled={hasReadAccess || cancelSubscriptionLoading || subscriptionsLoading}
    >
      Unsubscribe
    </Button>
  ) : (
    <Button
      onClick={createSubscriptionHander(createSubscription, {
        developerId,
        applicationId: appId,
        user: email,
        type: subscriptionType,
      })}
      intent="primary"
      loading={createSubscriptionLoading || subscriptionsLoading}
      disabled={hasReadAccess || createSubscriptionLoading || subscriptionsLoading}
    >
      Subscribe
    </Button>
  )
}
