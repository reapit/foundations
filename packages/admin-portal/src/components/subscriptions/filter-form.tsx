import React, { Dispatch, FC, SetStateAction } from 'react'
import { useForm } from 'react-hook-form'
import { FormLayout, InputWrap, InputGroup, Label, Select } from '@reapit/elements'
import { SubscriptionsFilters } from '.'
import debounce from 'just-debounce-it'

export interface FilterFormProps {
  setSubscriptionsFilters: Dispatch<SetStateAction<SubscriptionsFilters>>
}

export const FilterForm: FC<FilterFormProps> = ({ setSubscriptionsFilters }) => {
  const { register, handleSubmit } = useForm<SubscriptionsFilters>({ mode: 'all' })

  return (
    <form onChange={handleSubmit(debounce(setSubscriptionsFilters, 500))}>
      <FormLayout hasMargin>
        <InputWrap>
          <InputGroup {...register('organisationName')} label="Company Name" />
        </InputWrap>
        <InputWrap>
          <InputGroup {...register('userEmail')} type="email" label="User Email" />
        </InputWrap>
        <InputWrap>
          <InputGroup>
            <Label>Subscription Type</Label>
            <Select {...register('subscriptionType')}>
              <option value="">Please Select</option>
              <option value="applicationListing">Application Listing</option>
              <option value="developerRegistration">Developer Registration</option>
              <option value="developerEdition">Developer Edition</option>
            </Select>
          </InputGroup>
        </InputWrap>
        <InputWrap>
          <InputGroup>
            <Label>Status</Label>
            <Select {...register('status')}>
              <option value="">Please Select</option>
              <option value="active">Active</option>
              <option value="cancelled">Cancelled</option>
            </Select>
          </InputGroup>
        </InputWrap>
      </FormLayout>
    </form>
  )
}
