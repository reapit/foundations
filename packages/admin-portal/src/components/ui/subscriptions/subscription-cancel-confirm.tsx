import React, { useEffect } from 'react'
import { Button, Section, ModalV2, H5 } from '@reapit/elements'
export type CancelConfirmModalProps = {
  onConfirm: () => void
  isShowConfirmModal: boolean
  setCancelSubId: React.Dispatch<React.SetStateAction<string>>
  isCanceling: boolean
}

export const CancelConfirmModal: React.FC<CancelConfirmModalProps> = ({
  onConfirm,
  isShowConfirmModal,
  setCancelSubId,
  isCanceling,
}) => {
  useEffect(() => {
    console.log(isCanceling)
    if (!isCanceling) {
      setCancelSubId('')
    }
  }, [isCanceling])

  const handleOnClose = () => setCancelSubId('')
  return (
    <ModalV2 visible={isShowConfirmModal} destroyOnClose={true} onClose={handleOnClose}>
      <H5>Are you sure you want to cancel this appointment?</H5>
      <Section isFlex hasPadding={false} hasMargin={false}>
        <Button variant="info" loading={isCanceling} onClick={onConfirm} type="button">
          Yes
        </Button>
        <Button variant="info" disabled={isCanceling} onClick={handleOnClose} type="button">
          No
        </Button>
      </Section>
    </ModalV2>
  )
}

export default CancelConfirmModal
