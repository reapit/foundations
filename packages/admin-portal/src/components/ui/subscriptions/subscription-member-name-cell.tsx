import React from 'react'
import { useSelector } from 'react-redux'
import { selectDeveloperListState } from '@/selector/admin'

export interface MemberNameCellProps {
  cell: { value: string }
}

const MemberNameCell: React.FC<MemberNameCellProps> = ({ cell: { value } }) => {
  const DeveloperListState = useSelector(selectDeveloperListState)
  const { data } = DeveloperListState
  const developer = data.find(dev => dev.id === value)

  if (!developer) return null

  return <p>{developer.name}</p>
}

export default MemberNameCell
