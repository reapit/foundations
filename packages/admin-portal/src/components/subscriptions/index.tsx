import React, { Dispatch, FC, SetStateAction, useEffect, useState } from 'react'
import ErrorBoundary from '../error-boundary'
import { toLocalTime } from '@reapit/utils-common'
import {
  PageContainer,
  Loader,
  Title,
  Pagination,
  Table,
  elMb11,
  useModal,
  BodyText,
  ButtonGroup,
  Button,
} from '@reapit/elements'
import {
  objectToQuery,
  SendFunction,
  useReapitGet,
  useReapitUpdate,
  UpdateActionNames,
  updateActions,
  GetActionNames,
  getActions,
} from '@reapit/use-reapit-data'
import { SubscriptionModelPagedResult } from '@reapit/foundations-ts-definitions'
import { reapitConnectBrowserSession } from '../../core/connect-session'
import { FilterForm } from './filter-form'
import { Statistics } from '../statistics'
import { usePermissionsState } from '../../core/use-permissions-state'

export interface SubscriptionsFilters {
  subscriptionType?: string
  organisationName?: string
  userEmail?: string
  status?: string
}

export const handleCancelSub =
  (cancelSub: SendFunction<void, boolean | null>, setCancelSubId: Dispatch<SetStateAction<string | null>>) =>
  async () => {
    const cancelled = await cancelSub()
    if (cancelled) {
      setCancelSubId(null)
    }
  }

export const handleCancelSubSuccess = (refetchSubs: () => void, closeModal: () => void, success?: boolean) => () => {
  if (success) {
    refetchSubs()
    closeModal()
  }
}

export const handleSetSubId =
  (setCancelSubId: Dispatch<SetStateAction<string | null>>, openModal: () => void, cancelSubId?: string) => () => {
    if (cancelSubId) {
      openModal()
      setCancelSubId(cancelSubId)
    }
  }

const Subscriptions: FC = () => {
  const [subscriptionsFilters, setSubscriptionsFilters] = useState<SubscriptionsFilters>({})
  const [pageNumber, setPageNumber] = useState<number>(1)
  const [pageSize, setPageSize] = useState<number>(12)
  const [cancelSubId, setCancelSubId] = useState<string | null>(null)
  const { Modal, openModal, closeModal } = useModal()
  const { hasReadAccess } = usePermissionsState()

  const [subscriptions, subscriptionsLoading, , refetchSubs] = useReapitGet<SubscriptionModelPagedResult>({
    reapitConnectBrowserSession,
    action: getActions[GetActionNames.getSubscriptions],
    queryParams: {
      ...objectToQuery(subscriptionsFilters),
      userEmail: subscriptionsFilters.userEmail ? encodeURIComponent(subscriptionsFilters.userEmail ?? '') : undefined,
      pageSize,
      pageNumber,
    },
  })

  const [, , cancelSub, cancelSubSuccess] = useReapitUpdate<void, null>({
    reapitConnectBrowserSession,
    action: updateActions[UpdateActionNames.deleteSubscription],
    method: 'DELETE',
    uriParams: {
      subscriptionId: cancelSubId,
    },
  })

  useEffect(handleCancelSubSuccess(refetchSubs, closeModal, cancelSubSuccess), [cancelSubSuccess])

  return (
    <ErrorBoundary>
      <PageContainer>
        <Title>Subscriptions</Title>
        <FilterForm setSubscriptionsFilters={setSubscriptionsFilters} />
        <Statistics area="SUBSCRIPTIONS" data={subscriptions} setPageSize={setPageSize} />
        {subscriptionsLoading ? (
          <Loader />
        ) : (
          <>
            <Table
              className={elMb11}
              numberColumns={10}
              rows={subscriptions?.data?.map(
                ({ id, type, summary, organisationName, user, created, renews, frequency, cost, cancelled }) => ({
                  cells: [
                    {
                      label: 'Subcription Type',
                      value: type,
                      cellHasDarkText: true,
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Summary',
                      value: summary,
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Company Name',
                      value: organisationName,
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'User Email',
                      value: user,
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Start Date',
                      value: toLocalTime(created),
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Renews',
                      value: renews ? toLocalTime(renews) : '-',
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Frequency',
                      value: frequency,
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Cost',
                      value: `Â£${cost}`,
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                    {
                      label: 'Cancelled',
                      value: cancelled ? toLocalTime(cancelled) : '-',
                      narrowTable: {
                        showLabel: true,
                      },
                    },
                  ],
                  ctaContent: {
                    headerContent: 'Cancel',
                    icon: cancelled || hasReadAccess ? undefined : 'trashSystem',
                    onClick: cancelled || hasReadAccess ? undefined : handleSetSubId(setCancelSubId, openModal, id),
                  },
                }),
              )}
            />
            <Pagination
              callback={setPageNumber}
              currentPage={pageNumber}
              numberPages={Math.ceil((subscriptions?.totalCount ?? 1) / 12)}
            />
            <Modal title="Cancel Subscription">
              <BodyText>Please confirm you wish to cancel this subscription</BodyText>
              <ButtonGroup alignment="right">
                <Button type="button" onClick={closeModal}>
                  Close
                </Button>
                <Button intent="danger" onClick={handleCancelSub(cancelSub, setCancelSubId)}>
                  Cancel
                </Button>
              </ButtonGroup>
            </Modal>
          </>
        )}
      </PageContainer>
    </ErrorBoundary>
  )
}

export default Subscriptions
