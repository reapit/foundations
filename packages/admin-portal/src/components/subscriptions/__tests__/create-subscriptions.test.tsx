import { SubscriptionModel } from '@reapit/foundations-ts-definitions'
import { useReapitGet } from '@reapit/utils-react'
import React from 'react'
import { render } from '../../../tests/react-testing'
import { mockSubscriptionModelPagedResult } from '../../../tests/__stubs__/subscriptions'
import {
  cancelSubscriptionHander,
  createSubscriptionHander,
  CreateSubscriptions,
  CreateSubscriptionsProps,
  handleFetchSubs,
  handleUpdateAction,
  ToggleSubscribedForm,
} from '../create-subscriptions'

jest.mock('../../../core/use-permissions-state')
jest.mock('@reapit/utils-react', () => ({
  useReapitGet: jest.fn(() => [null, false, undefined, jest.fn()]),
  useReapitUpdate: jest.fn(() => [false, undefined, jest.fn(), true]),
  objectToQuery: jest.fn(),
}))

const mockUseReapitGet = useReapitGet as jest.Mock

const props: CreateSubscriptionsProps = {
  subscriptionType: 'applicationListing',
  developerId: 'DEVELOPER_ID',
  appId: 'APP_ID',
}

describe('CreateSubscriptions', () => {
  it('should match a snapshot with subscriptions and applicationListing', () => {
    mockUseReapitGet.mockReturnValue([mockSubscriptionModelPagedResult, false, undefined, jest.fn()])
    expect(render(<CreateSubscriptions {...props} />)).toMatchSnapshot()
  })

  it('should match a snapshot with subscriptions and developerRegistration', () => {
    mockUseReapitGet.mockReturnValue([mockSubscriptionModelPagedResult, false, undefined, jest.fn()])
    expect(
      render(<CreateSubscriptions {...{ ...props, subscriptionType: 'developerRegistration' }} />),
    ).toMatchSnapshot()
  })
})

describe('createSubscriptionHander', () => {
  it('should handle creating subscriptions', () => {
    const createSubscription = jest.fn()
    const createSubscriptionModel = mockSubscriptionModelPagedResult.data as SubscriptionModel[][0]
    const curried = createSubscriptionHander(createSubscription, createSubscriptionModel)

    curried()

    expect(createSubscription).toHaveBeenCalledWith(createSubscriptionModel)
  })
})

describe('cancelSubscriptionHander', () => {
  it('should handle canceling subscriptions', () => {
    const cancelSubscription = jest.fn()
    const curried = cancelSubscriptionHander(cancelSubscription)

    curried()

    expect(cancelSubscription).toHaveBeenCalledTimes(1)
  })
})

describe('handleFetchSubs', () => {
  it('should handle fetching subscriptions', () => {
    const cancelSubscription = jest.fn()
    const curried = handleFetchSubs(cancelSubscription)

    curried()

    expect(cancelSubscription).toHaveBeenCalledTimes(1)
  })
})

describe('handleUpdateAction', () => {
  it('should handle cancelling a sub', () => {
    const createSub = jest.fn()
    const cancelSub = jest.fn()
    const formValues = {
      isSubscribed: 'NOT_SUBSCRIBED',
    } as ToggleSubscribedForm

    const curried = handleUpdateAction(createSub, cancelSub)

    curried(formValues)

    expect(cancelSub).toHaveBeenCalledTimes(1)
    expect(createSub).not.toHaveBeenCalled()
  })

  it('should handle creating a sub', () => {
    const createSub = jest.fn()
    const cancelSub = jest.fn()
    const formValues = {
      isSubscribed: 'SUBSCRIBED',
    } as ToggleSubscribedForm

    const curried = handleUpdateAction(createSub, cancelSub)

    curried(formValues)

    expect(cancelSub).not.toHaveBeenCalled()
    expect(createSub).toHaveBeenCalledTimes(1)
  })
})
