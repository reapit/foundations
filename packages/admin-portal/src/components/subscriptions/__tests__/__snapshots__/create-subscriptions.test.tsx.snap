// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateSubscriptions should match a snapshot with subscriptions and applicationListing 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mocked-styled-93 el-snack-holder"
      />
      <form>
        <div
          class="el-mt5 el-form-layout"
        >
          <div
            class="mocked-styled-133 el-input-wrap"
          >
            <label
              class="mocked-styled-41 el-label"
            >
              Is Subscribed 
              Application Listing
            </label>
            <div
              class="el-has-grey-bg el-toggle-radio-wrap"
            >
              <input
                class="mocked-styled-127 el-toggle-radio"
                id="option-subscribed-true-MOCK_UUID"
                name="isSubscribed"
                type="radio"
                value="SUBSCRIBED"
              />
              <label
                class="el-has-grey-bg el-toggle-radio-label"
                for="option-subscribed-true-MOCK_UUID"
                role="button"
                tabindex="0"
              >
                <span
                  class="mocked-styled-124 el-toggle-radio-item"
                >
                  Subscribed
                </span>
              </label>
              <input
                checked=""
                class="mocked-styled-127 el-toggle-radio"
                id="option-subscribed-false-MOCK_UUID"
                name="isSubscribed"
                type="radio"
                value="NOT_SUBSCRIBED"
              />
              <label
                class="el-has-grey-bg el-toggle-radio-label"
                for="option-subscribed-false-MOCK_UUID"
                role="button"
                tabindex="0"
              >
                <span
                  class="mocked-styled-124 el-toggle-radio-item"
                >
                  Not Subscribed
                </span>
              </label>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <div
      class="mocked-styled-93 el-snack-holder"
    />
    <form>
      <div
        class="el-mt5 el-form-layout"
      >
        <div
          class="mocked-styled-133 el-input-wrap"
        >
          <label
            class="mocked-styled-41 el-label"
          >
            Is Subscribed 
            Application Listing
          </label>
          <div
            class="el-has-grey-bg el-toggle-radio-wrap"
          >
            <input
              class="mocked-styled-127 el-toggle-radio"
              id="option-subscribed-true-MOCK_UUID"
              name="isSubscribed"
              type="radio"
              value="SUBSCRIBED"
            />
            <label
              class="el-has-grey-bg el-toggle-radio-label"
              for="option-subscribed-true-MOCK_UUID"
              role="button"
              tabindex="0"
            >
              <span
                class="mocked-styled-124 el-toggle-radio-item"
              >
                Subscribed
              </span>
            </label>
            <input
              checked=""
              class="mocked-styled-127 el-toggle-radio"
              id="option-subscribed-false-MOCK_UUID"
              name="isSubscribed"
              type="radio"
              value="NOT_SUBSCRIBED"
            />
            <label
              class="el-has-grey-bg el-toggle-radio-label"
              for="option-subscribed-false-MOCK_UUID"
              role="button"
              tabindex="0"
            >
              <span
                class="mocked-styled-124 el-toggle-radio-item"
              >
                Not Subscribed
              </span>
            </label>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`CreateSubscriptions should match a snapshot with subscriptions and developerRegistration 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mocked-styled-93 el-snack-holder"
      />
      <form>
        <div
          class="el-mt5 el-form-layout"
        >
          <div
            class="mocked-styled-133 el-input-wrap"
          >
            <label
              class="mocked-styled-41 el-label"
            >
              Is Subscribed 
              Developer Registration
            </label>
            <div
              class="el-has-grey-bg el-toggle-radio-wrap"
            >
              <input
                checked=""
                class="mocked-styled-127 el-toggle-radio"
                id="option-subscribed-true-MOCK_UUID"
                name="isSubscribed"
                type="radio"
                value="SUBSCRIBED"
              />
              <label
                class="el-has-grey-bg el-toggle-radio-label"
                for="option-subscribed-true-MOCK_UUID"
                role="button"
                tabindex="0"
              >
                <span
                  class="mocked-styled-124 el-toggle-radio-item"
                >
                  Subscribed
                </span>
              </label>
              <input
                class="mocked-styled-127 el-toggle-radio"
                id="option-subscribed-false-MOCK_UUID"
                name="isSubscribed"
                type="radio"
                value="NOT_SUBSCRIBED"
              />
              <label
                class="el-has-grey-bg el-toggle-radio-label"
                for="option-subscribed-false-MOCK_UUID"
                role="button"
                tabindex="0"
              >
                <span
                  class="mocked-styled-124 el-toggle-radio-item"
                >
                  Not Subscribed
                </span>
              </label>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <div
      class="mocked-styled-93 el-snack-holder"
    />
    <form>
      <div
        class="el-mt5 el-form-layout"
      >
        <div
          class="mocked-styled-133 el-input-wrap"
        >
          <label
            class="mocked-styled-41 el-label"
          >
            Is Subscribed 
            Developer Registration
          </label>
          <div
            class="el-has-grey-bg el-toggle-radio-wrap"
          >
            <input
              checked=""
              class="mocked-styled-127 el-toggle-radio"
              id="option-subscribed-true-MOCK_UUID"
              name="isSubscribed"
              type="radio"
              value="SUBSCRIBED"
            />
            <label
              class="el-has-grey-bg el-toggle-radio-label"
              for="option-subscribed-true-MOCK_UUID"
              role="button"
              tabindex="0"
            >
              <span
                class="mocked-styled-124 el-toggle-radio-item"
              >
                Subscribed
              </span>
            </label>
            <input
              class="mocked-styled-127 el-toggle-radio"
              id="option-subscribed-false-MOCK_UUID"
              name="isSubscribed"
              type="radio"
              value="NOT_SUBSCRIBED"
            />
            <label
              class="el-has-grey-bg el-toggle-radio-label"
              for="option-subscribed-false-MOCK_UUID"
              role="button"
              tabindex="0"
            >
              <span
                class="mocked-styled-124 el-toggle-radio-item"
              >
                Not Subscribed
              </span>
            </label>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
